src/backend/distributed/transaction/relation_access_tracking.c

```diff
@@ -281,11 +284,68 @@ RecordPlacementAccessToCache(Oid relationId, ShardPlacementAccessType accessType
 }
 
 
+/*
+ * RecordParallelAccessForTaskList gets a task list and records
+ * the necessary parallel relation accesses for the task list.
+ *
+ * This function is used to enforce foreign keys from distributed
+ * tables to reference tables.
+ */
+void
+RecordParallelAccessForTaskList(List *taskList)
+{
+	Task *firstTask = NULL;
+
+	if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+	{
+		/* sequential mode prevents parallel access */
+		return;
+	}
+
+	if (list_length(taskList) < 2)
+	{
+		/* single shard task doesn't mean parallel access in our definition */
+		return;
+	}
+
+	firstTask = linitial(taskList);
+
+	if (firstTask->taskType == SQL_TASK)
+	{
+		RecordRelationParallelSelectAccessForTask(firstTask);
+	}
+	else if (firstTask->taskType == MODIFY_TASK)
+	{
+		if (firstTask->rowValuesLists != NIL)
+		{
+			/*
+			 * We always run multi-row INSERTs in a sequential
+			 * mode (hard-coded). Thus, we do not mark as parallel
+			 * access even if the prerequisites hold.
+			 */
+		}
+		else
+		{
+			/*
+			 * We prefer to mark with all remaining multi-shard modifications
+			 * with both modify and select accesses.
+			 */
+			RecordRelationParallelModifyAccessForTask(firstTask);
+			RecordRelationParallelSelectAccessForTask(firstTask);
+		}
+	}
+	else
+	{
+		RecordRelationParallelDDLAccessForTask(firstTask);
```

Comment:
In the previous version of the code we are skipping this for reference tables. Why are we not doing that anymore?

Reply:
I made some clean-up. In the previous code, we've the following comment, which is actually true:
```
		/*	
		 * Even single task DDLs hit here, so we'd prefer	
		 * not to record for reference tables.	
		 */
```

But, with this function we return early for that case `if (list_length(taskList) < 2)` which is also easier to read for me. What do you think?